### PR TITLE
feat(sidecar): min priority fee

### LINKED ISSUE #228: Minimum priority fee config
## Context
Right now we accept all commitment requests. We should probably set up a configuration option for a minimum priority fee.

### PR DESCRIPTION
**Pull request by namn-grg**
fixes #228 

## Release Notes

- **New Features**
	- Enhanced configuration options for setting minimum priority fees via command-line arguments.
	- Added validation methods to ensure priority fees meet specified conditions.

- **Bug Fixes**
	- Improved validation logic for transaction fees, ensuring better handling of maximum and minimum priority fees.

- **Documentation**
	- Clarified documentation for the `calculate_max_basefee` function regarding input units.

- **Refactor**
	- Streamlined code in testing utilities for improved efficiency and clarity.

### CODE REVIEW

File: bolt-sidecar/src/config/mod.rs [hunk 1]
```diff
@@ -55,6 +55,9 @@ pub struct Opts {
     /// Max committed gas per slot
     #[clap(long, env = "BOLT_SIDECAR_MAX_COMMITTED_GAS")]
     pub(super) max_committed_gas: Option<NonZero<u64>>,
+    /// Min priority fee to accept for a commitment
+    #[clap(long, env = "BOLT_SIDECAR_MIN_PRIORITY_FEE")]
+    pub(super) min_priority_fee: Option<NonZero<u64>>,
```
Comment (thedevbirb) [inline bolt-sidecar/src/config/mod.rs:60]: We should probably add a default for this, like 0.5/1 gwei but let's discuss it a bit

File: bolt-sidecar/src/primitives/commitment.rs [hunk 2]
```diff
@@ -156,6 +156,19 @@ impl InclusionRequest {
         true
     }
 
+    pub fn validate_min_priority_fee(&self, min_priority_fee: u128) -> bool {
+        for tx in &self.txs {
+            if tx
+                .max_priority_fee_per_gas()
+                .is_some_and(|max_priority_fee| max_priority_fee < min_priority_fee)
```
Comment (thedevbirb) [inline bolt-sidecar/src/primitives/commitment.rs]: What if in all the `txs` the `max_priority_fee_per_gas` is `None`? That is the case for legacy transactions and should be handled carefully by looking at the `gas_price` instead.

File: bolt-sidecar/src/config/mod.rs (hunk 1 above)
Comment (thedevbirb) [inline bolt-sidecar/src/config/mod.rs]: Isn't `u128` a better type to express priority fees? I'm not sure tho, could you take a look?

File: bolt-sidecar/src/state/execution.rs [hunk 3]
```diff
@@ -810,6 +819,41 @@ mod tests {
         Ok(())
     }
 
+    #[tokio::test]
+    async fn test_invalid_inclusion_request_min_priority_fee() -> eyre::Result<()> {
```
Comment (thedevbirb) [inline bolt-sidecar/src/state/execution.rs:831]: Could you please also do the test with legacy transactions?

**Review by thedevbirb — COMMENTED** [review]
Thanks for the PR! Some observations

File: bolt-sidecar/src/config/mod.rs [hunk 4]
```diff
@@ -158,13 +161,17 @@ pub struct Limits {
...
+    /// Minimum priority fee to accept for a commitment
+    pub min_priority_fee: NonZero<u64>,
 }
 
 impl Default for Limits {
     fn default() -> Self {
         Self {
             max_commitments_per_slot: NonZero::new(128).expect("Valid non-zero"),
             max_committed_gas_per_slot: NonZero::new(10_000_000).expect("Valid non-zero"),
+            min_priority_fee: NonZero::new(2).expect("Valid non-zero"),
```
Comment (thedevbirb) [inline bolt-sidecar/src/config/mod.rs]: Can you expose those values in constants? I think it is easier to notice them in that way. Thanks!

File: bolt-sidecar/src/primitives/commitment.rs [hunk 5]
```diff
@@ -143,12 +143,23 @@ impl InclusionRequest {
...
         for tx in &self.txs {
-            if tx
-                .max_priority_fee_per_gas()
-                .is_some_and(|max_priority_fee| max_priority_fee > tx.max_fee_per_gas())
-            {
+            let gas_price = tx.as_legacy().map(|tx| tx.gas_price).unwrap_or(0);
+            let max_priority_fee = tx.max_priority_fee_per_gas().unwrap_or(0);
+            if max_priority_fee + gas_price > tx.max_fee_per_gas() {
+                return false;
+            }
```
Comment (merklefruit) [inline bolt-sidecar/src/primitives/commitment.rs:152]: I don't fully get this, what does `tx.max_fee_per_gas` have to do with this?

Comment (namn-grg) [inline bolt-sidecar/src/primitives/commitment.rs]: btw this was already there though

Comment (merklefruit) [reply bolt-sidecar/src/primitives/commitment.rs]: oops I removed it, yeah I see it was already present but I still am not sure about the usage of `gas_price` here

Comment (merklefruit) [reply bolt-sidecar/src/primitives/commitment.rs]: The above code (that was already there) doesn't make sense to me, trying to understand it more

File: bolt-sidecar/src/primitives/commitment.rs [hunk 6]
```diff
@@ -143,12 +146,34 @@ impl InclusionRequest {
...
+            let gas_price = tx.as_legacy().map(|tx| tx.gas_price).unwrap_or(0);
+            let gas_price = if max_base_fee >= gas_price {
+                0
+            } else {
+                (gas_price - max_base_fee) / GWEI_TO_WEI as u128
+            };
+            let max_priority_fee = tx.max_priority_fee_per_gas().unwrap_or(0);
+
+            if gas_price + max_priority_fee < min_priority_fee {
                 return false;
```
Comment (thedevbirb) [inline bolt-sidecar/src/primitives/commitment.rs]: I don't have clear some points:
- How do you handle type 1 / EIP-2930 transactions?
- Why are you dividing the gas price for `GWEI_TO_WEI`?
- Can't you just check if to use `gas_price` or `max_priority_fee` by checking the type of the transaction?
- Can you add some comments as well with steps and motivation?

Comment (namn-grg) [reply bolt-sidecar/src/primitives/commitment.rs]: good point, addressed, thanks

File: bolt-sidecar/src/config/mod.rs [hunk 7]
```diff
@@ -28,6 +28,11 @@ pub const DEFAULT_RPC_PORT: u16 = 8000;
 /// Default port for the Constraints proxy server.
 pub const DEFAULT_CONSTRAINTS_PROXY_PORT: u16 = 18551;
 
+// Default limit values
+pub const DEFAULT_MAX_COMMITMENTS: usize = 128;
+pub const DEFAULT_MAX_COMMITTED_GAS: u64 = 10_000_000;
+pub const DEFAULT_MIN_PRIORITY_FEE: u128 = 1000000000; // 1 Gwei
```
Comment (thedevbirb) [inline bolt-sidecar/src/config/mod.rs:34]:
```suggestion
pub const DEFAULT_MIN_PRIORITY_FEE: u128 = 1_000_000_000; // 1 Gwei
```

File: bolt-sidecar/src/config/mod.rs [hunk 8]
```diff
@@ -55,6 +60,9 @@ pub struct Opts {
     /// Max committed gas per slot
     #[clap(long, env = "BOLT_SIDECAR_MAX_COMMITTED_GAS")]
     pub(super) max_committed_gas: Option<NonZero<u64>>,
+    /// Min priority fee to accept for a commitment
```
Comment (thedevbirb) [inline bolt-sidecar/src/config/mod.rs:63]:
```suggestion
    /// Min priority fee to accept for a commitment, in wei
```

File: bolt-sidecar/src/primitives/commitment.rs [hunk 9]
```diff
@@ -143,16 +144,67 @@ impl InclusionRequest {
...
+                        TxType::Legacy => tx.as_legacy().map(|tx| tx.gas_price),
+                        TxType::Eip2930 => tx.as_eip2930().map(|tx| tx.gas_price),
+                        _ => None,
+                    }
+                    .unwrap_or(0)
+                }
+                // For EIP-1559 (or any other), the effective priority fee is the max priority fee
+                // per gas
+                _ => tx.max_priority_fee_per_gas().unwrap_or(0),
+            };
```
Comment (thedevbirb) [inline bolt-sidecar/src/primitives/commitment.rs]:
```suggestion
            let effective_fee = if tx.tx_type() == TxType::Legacy {
                tx.as_legacy().expect("legacy tx").gas_price
            } else if tx.tx_type() == TxType::Eip2930 {
                tx.as_eip2930().expect("Eip2930 tx").gas_price
            } else {
                // For EIP-1559 (or any other), the effective priority fee 
                // is the max priority fee per gas
                tx.max_priority_fee_per_gas().expect("Post EIP1559 tx")
            };

```
This is more a styling suggestion but lmk if you like it! I think it is more succint and having some good asserts can be helpful.

File: bolt-sidecar/src/primitives/commitment.rs [hunk 10]
```diff
@@ -143,16 +144,67 @@ impl InclusionRequest {
...
+
+                    // Compute the priority fee by subtracting the base fee from the gas price.
+                    // If the gas price is less than or equal to the base fee, the priority fee is
+                    // 0.
+                    gas_price.saturating_sub(max_base_fee)
+                }
+                // For EIP-1559 (or any other), the effective priority fee is the max priority fee
+                // per gas
+                _ => tx.max_priority_fee_per_gas().unwrap_or(0),
+            };
```
Comment (thedevbirb) [inline bolt-sidecar/src/primitives/commitment.rs]:
```suggestion
            // For legacy/EIP-2930 transactions, compute the priority fee by subtracting the base fee from the gas price.
            // If the gas price is less than or equal to the base fee, the priority fee is 0.
            let effective_fee = if tx.tx_type() == TxType::Legacy {
                tx.as_legacy().expect("legacy tx").gas_price.saturating_sub(max_base_fee)
            } else if tx.tx_type() == TxType::Eip2930 {
                tx.as_eip2930().expect("Eip2930 tx").gas_price.saturating_sub(max_base_fee)
            } else {
                // For EIP-1559 (or any other), the effective priority fee 
                // is the max priority fee per gas
                tx.max_priority_fee_per_gas().expect("Post EIP1559 tx")
            };

```

In the same fashion of the previous comment. Lmk if you like it more!

**Review by thedevbirb — COMMENTED** [review]
Some nits! Now it looks good

File: bolt-sidecar/src/state/execution.rs (hunk 3 above)
Comment (thedevbirb) [reply bolt-sidecar/src/state/execution.rs:831]: Little bump on this: can you add an EIP-2930 tx too?

**Review by merklefruit — APPROVED** [review]
I like reusing Reth's logic, +1

**Comment by mempirate** [comment]
@coderabbitai review

File: bolt-sidecar/src/primitives/commitment.rs [hunk 11]
```diff
@@ -143,19 +143,39 @@ impl InclusionRequest {
...
+
+            // Check if the effective gas tip is more than the minimum priority fee
+            if tip < min_priority_fee {
+                return false;
+            }
+        }
+
+        // If all transactions meet the minimum priority fee, return true
+        true
+    }
```
[the remark this answers is not in the document]
Comment (merklefruit) [reply bolt-sidecar/src/primitives/commitment.rs]: this is nice

**Review by thedevbirb — APPROVED** [review]
Agreed that coderabbit made some nice suggestions! If you include them this is good

**Pull request merged**